index.js

```diff
@@ -25,27 +25,52 @@ function compose (middleware) {
   }
 
   /**
+   * Make an iterator for middleware.
+   */
+
+  middleware[Symbol.iterator] = function () {
+    const self = this
+    let i = 0
+    let context
+    let nextFunc
+
+    return {
+      next (c, n) {
+        if (!context) context = c
+        if (!nextFunc) nextFunc = n
+
+        let fn = self[i++]
+        let done = i > self.length
+        let value
+
+        if (done) {
+          value = nextFunc ? nextFunc() : void 0
+        } else {
+          value = fn.call(this, context, this.next.bind(this))
+        }
+
+        return {
+          value,
+          done
+        }
+      }
+    }
+  }
+
+  // Alternative iteration.
+  const iter = middleware[Symbol.iterator]()
+
+  /**
    * @param {Object} context
    * @return {Promise}
    * @api public
    */
 
   return function (context, next) {
-    // last called middleware #
-    let index = -1
-    return dispatch(0)
-    function dispatch (i) {
-      if (i <= index) return Promise.reject(new Error('next() called multiple times'))
-      index = i
-      const fn = middleware[i] || next
-      if (!fn) return Promise.resolve()
-      try {
-        return Promise.resolve(fn(context, function next () {
-          return dispatch(i + 1)
-        }))
-      } catch (err) {
-        return Promise.reject(err)
-      }
-    }
+    return new Promise((resolve, reject) => {
+      let value = iter.next(context, next).value
+      if (value && value.then) value.then(resolve).catch(reject)
+      else resolve(value)
```

Comment:
I think this logic is overly complicated. Try doing the logic outside of a promise, and returning the resultant promise.